### PR TITLE
Improve HTTP Error Handling in R11s Driver

### PR DESCRIPTION
**Pull request by znewton**
Many of the errors in from r11s-driver are thrown as either a status code or a raw AxiosError. Following ODSP's example, this change transcribes request errors into more meaningful DriverErrors with `canRetry` and `errorType` properties. This will allow RetriableDocumentStorageService in the container layer to more accurately know when to retry a request. In addition, this adds a single retry with a re-generated auth token on 401 errors.

There will be a follow-up PR once the changes in #5116 are released in `@fluidframework/server-services-client@0.1020.0` to handle auth token refresh/retry logic in DocumentStorageService. For now, we can say `canRetry: false` for those errors because without token refresh it would just fail again on future retries.

### CODE REVIEW

File: packages/drivers/routerlicious-driver/src/requestUtils.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+                ? `${axiosError.response?.data} (${axiosError.response?.statusText})`
+                : axiosError.response?.data,
+            axiosError.response?.status,
+            isStatusRetriable(axiosError.response?.status, axiosError.response?.data?.retryAfter),
+            axiosError.response?.data?.retryAfter
+                ? axiosError.response?.data?.retryAfter
+                : undefined,
+        );
+    }
+    throwR11sNetworkError(`${error}`);
```
Comment (tanviraumi) [inline packages/drivers/routerlicious-driver/src/requestUtils.ts:23]: Is it guaranteed that error will always be axios error? If its not, is it always a string?

File: packages/drivers/routerlicious-driver/src/documentStorageService.ts [hunk 2]
```diff
@@ -225,9 +230,22 @@ export class DocumentStorageService implements IDocumentStorageService {
...
+            const blob = await this.manager.createBlob(parsedContent, encoding)
+                .catch(DocumentStorageService.enhanceGitServiceError);
             assert(hash === blob.sha, "Blob.sha and hash do not match!!");
         }
         return hash;
     }
+
+    public static async enhanceGitServiceError(error: any): Promise<never> {
+        const messageFallback = "GitManager call failed";
+        // GitManager's Historian's RestWrapper only throws status code when response is available
```
Comment (tanviraumi) [inline packages/drivers/routerlicious-driver/src/documentStorageService.ts:242]: Do we need to make historians error response better?

**Review by tanviraumi — APPROVED** [review]

**Comment by tanviraumi** [comment]
@vladsud Any input?

File: packages/drivers/routerlicious-driver/src/documentStorageService.ts (hunk 2 above)
Comment (znewton) [reply packages/drivers/routerlicious-driver/src/documentStorageService.ts:242]: Yes, I think improving RestWrapper's error response would be helpful. We lose a lot of info when the errors are parsed down to just a status code

File: packages/drivers/routerlicious-driver/src/requestUtils.ts (hunk 1 above)
Comment (znewton) [reply packages/drivers/routerlicious-driver/src/requestUtils.ts:23]: I was unsure about this. In all the cases I'm trying to catch, yes, it should be an Axios error, but by the nature of it being an error it felt like I shouldn't assume it always would be. It would probably be safest to wrap this in json-stringify-safe

File: packages/drivers/routerlicious-driver/src/deltaStorageService.ts [hunk 3]
```diff
@@ -69,18 +69,19 @@ export class DeltaStorageService implements IDeltaStorageService {
...
         );
 
-        if (storageToken) {
-            headers.Authorization = `Basic ${storageToken.jwt}`;
-        }
-
-        const ops = await Axios.get<ISequencedDocumentMessage[]>(
-            `${this.url}?${query}`, { headers });
-
         if (this.logger) {
```
Comment (vladsud) [inline packages/drivers/routerlicious-driver/src/deltaStorageService.ts:85]: Unrelated to your change, but I'd suggest making logger required through all layers. One can always use Null logger in test / non-prod environments

File: packages/drivers/routerlicious-driver/src/documentStorageService.ts [hunk 4]
```diff
@@ -64,15 +66,17 @@ export class DocumentStorageService implements IDocumentStorageService {
...
+            .catch(DocumentStorageService.enhanceGitServiceError);
         const tree = buildHierarchy(rawTree, this.blobsShaCache);
 
         this._logTailSha = ".logTail" in tree.trees ? tree.trees[".logTail"].blobs.logTail : undefined;
         return tree;
     }
 
     public async getVersions(versionId: string, count: number): Promise<IVersion[]> {
-        const commits = await this.manager.getCommits(versionId ? versionId : this.id, count);
+        const commits = await this.manager.getCommits(versionId ? versionId : this.id, count)
```
Comment (vladsud) [inline packages/drivers/routerlicious-driver/src/documentStorageService.ts:78]: I'm not sure I like the way error handling is done. There is really nothing preventing IGitManager being implemented differently (i.e. having alternative implementations) and have error reporting done in different way / structure. I feel like conversion to "right" format should happen as close to network call as possible, i.e. stay at the same layer.
Also not very clear where / how you are going to add retries for retryable errors (offline, 429, general errors, as well as expired token that requires token refresh).
Will it be at same layer? At IGitManager  layer?

File: packages/drivers/routerlicious-driver/src/r11sError.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    canRetry?: boolean,
+    retryAfterSeconds?: number,
+): R11sError {
+    switch (statusCode) {
+        case 401:
+        case 403:
+            return new NetworkErrorBasic(
+                errorMessage,
+                R11sErrorType.authorizationError,
+                canRetry ?? isStatusRetriable(statusCode, retryAfterSeconds),
```
Comment (vladsud) [inline packages/drivers/routerlicious-driver/src/r11sError.ts:40]: I'd inline isStatusRetriable() into each case - it's just easy to follow code

File: packages/drivers/routerlicious-driver/src/requestUtils.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+export async function enhanceRequestError(error: any): Promise<never> {
+    if (error?.isAxiosError) {
+        const axiosError = error as AxiosError;
+        throwR11sNetworkError(
+            axiosError.response?.statusText
+                ? `${axiosError.response?.data} (${axiosError.response?.statusText})`
+                : axiosError.response?.data,
+            axiosError.response?.status,
+            isStatusRetriable(axiosError.response?.status, axiosError.response?.data?.retryAfter),
```
Comment (vladsud) [inline packages/drivers/routerlicious-driver/src/requestUtils.ts:18]: I'd recommend removing canRetry argument - same isStatusRetriable() call is made inside throwR11sNetworkError.
I believe on ODSP side we eventually removed canRetry from similar flow as it was hard to follow code and reason on how to combine value of this flag with status code.
It's better in those cases where you want to overwrite it to just get error object from throwR11sNetworkError and overwrite canRetry property on object - easier to follow code and more obvious on what code does, without complicating throwR11sNetworkError

File: packages/drivers/routerlicious-driver/src/r11sError.ts [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+    // eslint-disable-next-line @typescript-eslint/no-throw-literal
+    throw networkError;
+}
+
+/**
+ * Assumes that 401 errors have already been retried with a new token, so further 401 retries will not succeed.
+ */
+export function isStatusRetriable(statusCode: number | undefined, retryAfterSeconds?: number) {
+    if (statusCode === undefined) {
```
Comment (vladsud) [inline packages/drivers/routerlicious-driver/src/r11sError.ts:82]: When can you have no status code?
Note - not having canRetry flag on error object is valid state, and we just switched DeltaManager to treat it as critical error (it was retryable for long time). We did it because we had a case where an assert in the code would result in infinite retry loop. And when we made that change in DeltaManager I found more cases like that.
In other words - if we have no clue where this error comes from and what it means, fail-fast, learn and address the issue and make code better for long term. Only retry what we know for sure is retryable.

File: packages/drivers/routerlicious-driver/src/r11sError.ts [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        return true;
+    }
+    if (statusCode === 429) {
+        return retryAfterSeconds !== undefined;
```
Comment (vladsud) [inline packages/drivers/routerlicious-driver/src/r11sError.ts:86]: I'd make it unconditional. If service returns 429 but does not tell me how long to wait, we will come back pretty quickly (though loader layer backs off in such cases and moves with a retry per 8 second at slowest pace).

**Comment by znewton** [comment]
@vladsud made some great points, so I opened #5302 to allow for a custom RestWrapper to be passed into Historian in DocumentStorageService. This will give us direct access to HTTP errors at the level the HTTP requests are made.

Closing this for now, because most of it will change if the Historian/RestWrapper changes go through.
